File: nodejs/consumer.js
// Copyright 2025 Padi, Inc. All Rights Reserved.

export class Consumer {
  #client;
  context;
  profile;

  constructor(client, context, profile) {
    this.#client = client;
    this.context = context;
    this.profile = profile;
  }

  watch(handler) {
    const keyPrefix = `cns/${this.context.node.system.id}/nodes/${this.context.node.id}/contexts/${this.context.id}/consumer/${this.profile}/`;
    const re = new RegExp(`connections/(\\w+)/properties/(\\w+)$`);
    this.#client.on('update', (event) => {
      for (let [key, value] of Object.entries(event.keys)) {
        if (!key.startsWith(keyPrefix)) {
          continue;
        }
        const captures = key.match(re);
        if (captures.length < 3) {
          continue;
        }
        const connection = captures[1];
        const property = captures[2];
        const changeEvent = {
          connection,
          property,
          value,
        };
        handler(changeEvent);
      }
    });
  }
}

File: python/src/arete_sdk/consumer.py
import re


class Consumer:
    def __init__(self, client, context, profile):
        self.client = client
        self.context = context
        self.profile = profile

    def profile_key_prefix(self):
        system_id = self.context.node.system.id
        node_id = self.context.node.id
        context_id = self.context.id
        profile = self.profile
        return f'cns/{system_id}/nodes/{node_id}/contexts/{context_id}/consumer/{profile}/'

    def property_key(self, property):
        profile_key_prefix = self.profile_key_prefix()
        return f'{profile_key_prefix}properties/{property}'

    def get(self, property):
        key = self.property_key(property)
        self.client.get(key)

    def put(self, property, value):
        key = self.property_key(property)
        self.client.put(key, value)

    def watch(self, fn):
        key_prefix = self.profile_key_prefix()

        # Start by notifying of existing cached properties
        for key, value in self.client.keys():
            if not key.startswith(key_prefix):
                continue
            captures = re.search('connections/(\\w+)/properties/(\\w+)$', key)
            connection = captures.group(1)
            property = captures.group(2)
            change_event = {
                'connection': connection,
                'property': property,
                'value': value,
            }
            fn(change_event)

        # Watch for future property changes
        def on_update(event):
            for key, value in event['keys'].items():
                if not key.startswith(key_prefix):
                    continue
                captures = re.search('connections/(\\w+)/properties/(\\w+)$', key)
                if captures is None:
                    continue
                connection = captures.group(1)
                property = captures.group(2)
                change_event = {
                    'connection': connection,
                    'property': property,
                    'value': value,
                }
                fn(change_event)
        self.client.on_update(on_update)
